src/class-convertkit-api.php

```diff
@@ -2036,66 +2290,103 @@ private function request( $endpoint, $method = 'get', $params = array(), $retry_
-					'convertkit_api_error',
-					$this->get_error_message( 'request_internal_server_error' ),
-					$http_response_code
-				);
 
-			// Bad gateway.
-			case 502:
-				return new WP_Error(
-					'convertkit_api_error',
-					$this->get_error_message( 'request_bad_gateway' ),
-					$http_response_code
-				);
-		}
+		// If the body is null i.e. a 204 No Content, don't attempt to JSON decode it.
+		$response = ( ! empty( $body ) ? json_decode( $body, true ) : null );
+
+		// Return a WP_Error if the HTTP response code is a 5xx code.
+		// The API respose won't contain an error message, so we use this class' error messages.
+		if ( $http_response_code >= 500 ) {
+			switch ( $http_response_code ) {
+				// Internal server error.
+				case 500:
+				default:
+					$error = $this->get_error_message( 'request_internal_server_error' );
+					break;
+
+				// Not implemented.
+				case 501:
+					$error = $this->get_error_message( 'request_method_unsupported' );
+					break;
+
+				// Bad gateway.
+				case 502:
+					$error = $this->get_error_message( 'request_bad_gateway' );
+					break;
+
+				// Service unavailable.
+				case 503:
+					$error = $this->get_error_message( 'request_service_unavailable' );
+					break;
+
+				// Gateway timeout.
+				case 504:
+					$error = $this->get_error_message( 'request_gateway_timeout' );
+					break;
+
+				// HTTP version not supported.
+				case 505:
+					$error = $this->get_error_message( 'request_http_not_supported' );
+					break;
+			}
 
-		// If the response is null for a non-DELETE method, json_decode() failed as the body could not be decoded.
-		if ( is_null( $response ) && $method !== 'delete' ) {
-			$this->log( 'API: Error: ' . sprintf( $this->get_error_message( 'response_type_unexpected' ), $body ) );
 			return new WP_Error(
 				'convertkit_api_error',
-				sprintf( $this->get_error_message( 'response_type_unexpected' ), $body ),
+				$error,
 				$http_response_code
 			);
 		}
 
-		// If an error message or code exists in the response, return a WP_Error.
-		if ( isset( $response['error'] ) ) {
-			// Build the message.
-			$message = $response['error'];
-			if ( array_key_exists( 'message', $response ) ) {
-				$message .= ': ' . $response['message'];
+		// Return the API error message as a WP_Error if the HTTP response code is a 4xx code.
+		if ( $http_response_code >= 400 ) {
+			// Define the error description.
+			$error = '';
+			if ( array_key_exists( 'errors', $response ) ) {
+				$error = implode( "\n", $response['errors'] );
+			} elseif ( array_key_exists( 'error_description', $response ) ) {
+				$error = $response['error_description'];
+			}
+
+			$this->log( 'API: Error: ' . $error );
+
+			switch ( $http_response_code ) {
+				// If the HTTP response code is 401, and the error matches 'The access token expired', refresh the access token now
+				// and re-attempt the request.
+				case 401:
+					if ( $error !== 'The access token expired' ) {
+						break;
```

Comment:
Would this suppress an error?

Reply:
No; it'll break out of the switch statement, meaning the next code executed will return a `WP_Error` object:
https://github.com/ConvertKit/convertkit-wordpress-libraries/pull/36/files/9a5c6a0a363a2da6a77b3fb4d1ce32d4eac69bf6#diff-f92d1b708df139635b79bf42702baa5c2f612a5a742fcc2793e40661cc71a6ceR2387